base_layer/wallet/src/output_manager_service/storage/database.rs

```diff
@@ -0,0 +1,376 @@
+// Copyright 2019. The Tari Project
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::output_manager_service::{error::OutputManagerStorageError, TxId};
+use chrono::{Duration as ChronoDuration, NaiveDateTime, Utc};
+use log::*;
+use std::{
+    collections::HashMap,
+    fmt::{Display, Error, Formatter},
+    time::Duration,
+};
+use tari_core::{tari_amount::MicroTari, transaction::UnblindedOutput, types::BlindingFactor};
+
+const LOG_TARGET: &'static str = "wallet::output_manager_service::database";
+
+/// This trait defines the required behaviour that a storage backend must provide for the Output Manager service.
+/// Data is passed to and from the backend via the [DbKey], [DbValue], and [DbValueKey] enums. If new data types are
+/// required to be supported by the backends then these enums can be updated to reflect this requirement and the trait
+/// will remain the same
+pub trait OutputManagerBackend {
+    /// Retrieve the record associated with the provided DbKey
+    fn fetch(&self, key: &DbKey) -> Result<Option<DbValue>, OutputManagerStorageError>;
+    /// Check if a record with the provided key exists in the backend.
+    fn contains(&self, key: &DbKey) -> Result<bool, OutputManagerStorageError>;
+    /// Modify the state the of the backend with a write operation
+    fn write(&mut self, op: WriteOperation) -> Result<Option<DbValue>, OutputManagerStorageError>;
+    /// This method must check if an output with the provided `spending_key` exists in any of the collections i.e.
```

Comment:
Careful. This trait is implementation-agnostic. You can't assume that such collections (in a data structure sense) exist. 
In a conceptual sense, yes; but being able to make that distinction is part of what makes this pattern so powerful and makes for a very clean abstraction.
E.g., in a SQL backend, all outputs might be in one table with a bool "spent" field. Then the `contains` implementation would be a very simple query.